File: ga/gpt3_5_result_2024_6/result/1968D_cpp_gpt/20_1968D.cpp
use std::io::{self, Write};

fn calculate(p: &Vec<usize>, a: &Vec<usize>, pos: usize, mut k: i64) -> i64 {
    let n = p.len();
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; n];
    let mut pos = pos;

    while !vis[pos] && k > 0 {
        vis[pos] = true;
        mx = mx.max(cur + (k * a[pos] as i64));
        cur += a[pos] as i64;
        pos = p[pos];
        k -= 1;
    }
    mx
}

fn main() {
    let stdout = io::stdout();
    let mut reader = io::BufReader::new(io::stdin());
    let mut input = String::new();
    reader.read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();

    for _ in 0..t {
        input.clear();
        reader.read_line(&mut input).unwrap();
        let mut iter = input.split_whitespace();
        let n: usize = iter.next().unwrap().parse().unwrap();
        let k: i64 = iter.next().unwrap().parse().unwrap();
        let ps1: usize = iter.next().unwrap().parse().unwrap();
        let ps2: usize = iter.next().unwrap().parse().unwrap();

        input.clear();
        reader.read_line(&mut input).unwrap();
        let p: Vec<usize> = input.split_whitespace()
            .map(|x| x.parse().unwrap())
            .map(|x: usize| x - 1)
            .collect();

        input.clear();
        reader.read_line(&mut input).unwrap();
        let a: Vec<usize> = input.split_whitespace().map(|x| x.parse().unwrap()).collect();

        let ans1 = calculate(&p, &a, ps1 - 1, k);
        let ans2 = calculate(&p, &a, ps2 - 1, k);

        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}

File: ga/4omini_result_2024_7/result/1968D_mini/20_1968D.rs
use std::cmp;
use std::io;

const N: usize = 200000;

fn maxl(a: i64, b: i64) -> i64 {
    if a > b { a } else { b }
}

fn main() {
    let stdin = io::stdin();
    let mut input = String::new();

    stdin.read_line(&mut input).unwrap();
    let mut iter = input.split_whitespace();
    let t: usize = iter.next().unwrap().parse().unwrap();

    for _ in 0..t {
        input.clear();
        stdin.read_line(&mut input).unwrap();
        let mut iter = input.split_whitespace();
        
        let n: usize = iter.next().unwrap().parse().unwrap();
        let k: usize = iter.next().unwrap().parse().unwrap();
        let mut a: usize = iter.next().unwrap().parse().unwrap();
        let mut b: usize = iter.next().unwrap().parse().unwrap();
        let m = cmp::min(n, k);
        
        let mut arr = vec![0; N + 1];
        let mut aa = vec![0; N + 1];

        input.clear();
        stdin.read_line(&mut input).unwrap();
        let iter = input.split_whitespace();
        for (i, value) in iter.enumerate() {
            arr[i + 1] = value.parse::<usize>().unwrap();
        }

        input.clear();
        stdin.read_line(&mut input).unwrap();
        let iter = input.split_whitespace();
        for (i, value) in iter.enumerate() {
            aa[i + 1] = value.parse::<i64>().unwrap();
        }
        
        let mut p1 = 0;
        let mut p2 = 0;
        let mut p1_ = 0;
        let mut p2_ = 0;
        
        for x in 1..=m {
            p1 = maxl(p1, aa[a] * (k as i64 - x as i64 + 1) + p1_);
            p1_ += aa[a];
            a = arr[a];
            
            p2 = maxl(p2, aa[b] * (k as i64 - x as i64 + 1) + p2_);
            p2_ += aa[b];
            b = arr[b];
        }
        
        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}